components/logins/src/encryption.rs

```diff
@@ -160,6 +174,136 @@ impl KeyManager for StaticKeyManager {
+///         // Most likely, you would want to prompt for a password.
+///         // let password = prompt_string("primary password").unwrap_or_default();
+///         Ok("secret".to_string())
+///     }
+///
+///     fn on_authentication_success(&self) {
+///         println!("success");
+///     }
+///
+///     fn on_authentication_failure(&self) {
+///         println!("this did not work, please try again:");
+///     }
+/// }
+/// let key_manager = NSSKeyManager::new(Arc::new(MyPrimaryPasswordAuthenticator {}));
+/// assert_eq!(key_manager.get_key().unwrap().len(), 63);
+/// ```
+#[cfg(feature = "keydb")]
+#[derive(uniffi::Object)]
+pub struct NSSKeyManager {
+    primary_password_authenticator: Arc<dyn PrimaryPasswordAuthenticator>,
+}
+
+#[cfg(feature = "keydb")]
+impl NSSKeyManager {
+    /// Initialize new `NSSKeyManager` with a given `PrimaryPasswordAuthenticator`.
+    /// There must be a previous initializiation of NSS before initializing
+    /// `NSSKeyManager`, otherwise this panics.
+    pub fn new(primary_password_authenticator: Arc<dyn PrimaryPasswordAuthenticator>) -> Self {
+        assert_nss_initialized();
+        Self {
+            primary_password_authenticator,
+        }
+    }
+}
+
+/// Identifier for the logins key, under which the key is stored in NSS.
+#[cfg(feature = "keydb")]
+static KEY_NAME: &str = "as-logins-key";
+
+// wrapp `authentication_with_primary_password_is_needed` into an ApiResult
+#[cfg(feature = "keydb")]
+fn api_authentication_with_primary_password_is_needed() -> ApiResult<bool> {
+    authentication_with_primary_password_is_needed().map_err(|e: nss::Error| {
+        LoginsApiError::NSSAuthenticationError {
+            reason: e.to_string(),
+        }
+    })
+}
+
+// wrapp `authenticate_with_primary_password` into an ApiResult
+#[cfg(feature = "keydb")]
+fn api_authenticate_with_primary_password(primary_password: &str) -> ApiResult<bool> {
+    authenticate_with_primary_password(primary_password).map_err(|e: nss::Error| {
+        LoginsApiError::NSSAuthenticationError {
+            reason: e.to_string(),
+        }
+    })
+}
+
+#[cfg(feature = "keydb")]
+impl KeyManager for NSSKeyManager {
+    fn get_key(&self) -> ApiResult<Vec<u8>> {
+        if api_authentication_with_primary_password_is_needed()? {
+            let primary_password =
+                block_on(self.primary_password_authenticator.get_primary_password())?;
+            let mut result = api_authenticate_with_primary_password(&primary_password)?;
+
+            if result {
+                self.primary_password_authenticator
+                    .on_authentication_success();
+            } else {
+                while !result {
+                    self.primary_password_authenticator
+                        .on_authentication_failure();
+
+                    let primary_password =
+                        block_on(self.primary_password_authenticator.get_primary_password())?;
```

Comment:
I wonder if later we want this to be async too and move the `block_on` to the caller, but this is fine for now.

Reply:
yes, good point. Would this be a breaking change for the key managers currently in use?

Reply:
Yes it would be.  However, are there any foreign implementations of `KeyManager` yet?  If not, then I don't think it will actually matter in practice and this a good time to sneak in this breaking change before there are any.

If there are, then you'll need to be a bit more careful.  If the only difference is that we're going to be changing `func` to `async func` then I think it's okay to introduce the breaking change and fix it when the next a-s nightly comes out.  We can work through that one together.

Either way, it's seems easier to make this change now than later, but I'll let you make that call.  We could also land this now, then make just that change in a follow-up.

Reply:
`KeyManager` is currently implemented in `StaticKeyManager`, which is used for Android and iOS. I absolutely agree that this should be asynchronous, but I'd like to make the change in a follow-up later this week to keep things short and simple.

Reply:
A follow-up is great.

`StaticKeyManager` is defined in the Rust code right?  I think that means you can make the change without having to change any of the consumer code, unless they're calling `get_key` themselves for some reason.  If all they're doing is creating a `StaticKeyManager` then passing it to the `LoginStore` constructor, then changing the method to be async should be no problem.